tsdb/wlog/wlog_test.go

```diff
@@ -208,6 +239,201 @@ func TestWALRepair_ReadingError(t *testing.T) {
 	}
 }
 
+// TestWALIgnoreCorruption ensures that if ignoreCorruption is set, reading will continue on remain valid records.
+func TestWALIgnoreCorruption(t *testing.T) {
+	for name, test := range map[string]struct {
+		corrSgm        int              // Which segment to corrupt.
+		corrFunc       func(f *os.File) // Func that applies the corruption.
+		intactRecs     int              // Total expected records left after the repair.
+		intactRecIndex []int            // Index of the intact record.
+	}{
+		"bad_fragment_sequence": {
+			1,
+			func(f *os.File) {
+				_, err := f.Seek(pageSize, 0)
+				require.NoError(t, err)
+				_, err = f.Write([]byte{byte(recLast)})
+				require.NoError(t, err)
+			},
+			8,
+			[]int{0, 1, 2, 3, 5, 6, 7, 8},
+		},
+		"bad_fragment_flag": {
+			1,
+			func(f *os.File) {
+				_, err := f.Seek(pageSize, 0)
+				require.NoError(t, err)
+				_, err = f.Write([]byte{123})
+				require.NoError(t, err)
+			},
+			8,
+			[]int{0, 1, 2, 3, 5, 6, 7, 8},
+		},
+		"bad_checksum": {
+			1,
+			func(f *os.File) {
+				_, err := f.Seek(pageSize+4, 0)
+				require.NoError(t, err)
+				_, err = f.Write([]byte{0})
+				require.NoError(t, err)
+			},
+			4,
+			[]int{0, 1, 2, 3},
+		},
+		"bad_length": {
+			1,
+			func(f *os.File) {
+				_, err := f.Seek(pageSize+2, 0)
+				require.NoError(t, err)
+				_, err = f.Write([]byte{0})
+				require.NoError(t, err)
+			},
+			4,
+			[]int{0, 1, 2, 3},
+		},
+		"bad_content": {
+			1,
+			func(f *os.File) {
+				_, err := f.Seek(pageSize+100, 0)
+				require.NoError(t, err)
+				_, err = f.Write([]byte("beef"))
+				require.NoError(t, err)
+			},
+			4,
+			[]int{0, 1, 2, 3},
+		},
+		"unexpected_middle": {
+			1,
+			func(f *os.File) {
+				_, err := f.Seek(pageSize, 0)
+				require.NoError(t, err)
+				_, err = f.Write([]byte{byte(recMiddle)})
+				require.NoError(t, err)
+			},
+			8,
+			[]int{0, 1, 2, 3, 5, 6, 7, 8},
+		},
+		"unexpected_full_after_first": {
+			1,
+			func(f *os.File) {
+				_, err := f.Seek(pageSize, 0)
+				require.NoError(t, err)
+				_, err = f.Write([]byte{byte(recFirst)})
+				require.NoError(t, err)
+			},
+			7,
+			[]int{0, 1, 2, 3, 6, 7, 8},
```

Comment:
I wonder in this case whether we should keep the full record, and just ignore the orphaned fragment preceding it? WDYT?